File: khalti-android/src/main/java/khalti/checkOut/EBanking/contactForm/ContactFormPresenter.java
package khalti.checkOut.EBanking.contactForm;

import android.support.annotation.NonNull;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.HashMap;

import khalti.checkOut.EBanking.helper.BankingData;
import khalti.checkOut.api.Config;
import khalti.utils.ApiUtil;
import khalti.utils.Constant;
import khalti.utils.EmptyUtil;
import khalti.utils.GuavaUtil;
import khalti.utils.NumberUtil;
import khalti.utils.StringUtil;
import khalti.utils.ValidationUtil;
import rx.subscriptions.CompositeSubscription;

class ContactFormPresenter implements ContactFormContract.Presenter {
    @NonNull
    private final ContactFormContract.View view;
    private CompositeSubscription compositeSubscription;

    ContactFormPresenter(@NonNull ContactFormContract.View view) {
        this.view = GuavaUtil.checkNotNull(view);
        view.setPresenter(this);
        compositeSubscription = new CompositeSubscription();
    }

    @Override
    public void onCreate() {
        BankingData bankingData = view.receiveData();
        if (EmptyUtil.isNotNull(bankingData)) {
            view.setBankData(bankingData.getBankLogo(), bankingData.getBankName(), bankingData.getBankIcon());
            view.setButtonText("Pay Rs " + StringUtil.formatNumber(NumberUtil.convertToRupees(bankingData.getConfig().getAmount())));
            compositeSubscription.add(view.setClickListener().subscribe(aVoid -> onFormSubmitted(view.isNetworkAvailable(), view.getContactNumber(), bankingData.getBankIdx(),
                    bankingData.getBankName(), bankingData.getConfig())));
            compositeSubscription.add(view.setEditTextListener().subscribe(charSequence -> view.setEditTextError(null)));
        }
    }

    @Override
    public void onDestroy() {
        if (EmptyUtil.isNotNull(compositeSubscription) && compositeSubscription.hasSubscriptions() && !compositeSubscription.isUnsubscribed()) {
            compositeSubscription.unsubscribe();
        }
    }

    @Override
    public void onFormSubmitted(boolean isNetwork, String mobile, String bankId, String bankName, Config config) {
        if (EmptyUtil.isNotEmpty(mobile) && ValidationUtil.isMobileNumberValid(mobile)) {
            if (isNetwork) {
                HashMap<String, Object> map = new HashMap<>();
                map.put("mobile", mobile);
                map.put("bankId", bankId);
                map.put("bankName", bankName);

                try {
                    String data = "public_key=" + URLEncoder.encode(config.getPublicKey(), "UTF-8") + "&" +
                            "product_identity=" + URLEncoder.encode(config.getProductId(), "UTF-8") + "&" +
                            "product_name=" + URLEncoder.encode(config.getProductName(), "UTF-8") + "&" +
                            "amount=" + URLEncoder.encode(config.getAmount() + "", "UTF-8") + "&" +
                            "mobile=" + URLEncoder.encode(map.get("mobile") + "", "UTF-8") + "&" +
                            "bank=" + URLEncoder.encode(map.get("bankId") + "", "UTF-8") + "&" +
                            "source=android" + "&" +
                            "return_url=" + URLEncoder.encode(view.getPackageName(), "UTF-8");

                    if (EmptyUtil.isNotNull(config.getProductUrl()) && EmptyUtil.isNotEmpty(config.getProductUrl())) {
                        data += "&" + "product_url=" + URLEncoder.encode(config.getProductUrl(), "UTF-8");
                    }

                    data += ApiUtil.getPostData(config.getAdditionalData());

                    view.saveConfigInFile(config);

                    view.openEBanking(Constant.url + "ebanking/initiate/?" + data);
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                    view.showMessageDialog("Error", "Something went wrong");
                }
            } else {
                view.showNetworkError();
            }
        } else {
            if (EmptyUtil.isEmpty(mobile)) {
                view.setEditTextError("This field is required");
            } else {
                view.setEditTextError("Enter a valid mobile number");
            }
        }
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/ebanking/contactForm/ContactFormPresenter.kt
package com.khalti.checkOut.ebanking.contactForm

import com.google.gson.Gson

import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.HashMap

import com.khalti.BuildConfig
import com.khalti.checkOut.ebanking.helper.BankingData
import com.khalti.checkOut.helper.Config
import com.khalti.signal.CompositeSignal
import com.khalti.utils.*
import com.khalti.utils.ErrorUtil.EMPTY_ERROR
import com.khalti.utils.ErrorUtil.MOBILE_ERROR
import rx.subscriptions.CompositeSubscription

internal class ContactFormPresenter(view: ContactFormContract.View) : ContactFormContract.Presenter {
    private val view: ContactFormContract.View = GuavaUtil.checkNotNull<ContactFormContract.View>(view)
    private val compositeSignal = CompositeSignal()

    init {
        view.setPresenter(this)
    }

    override fun onCreate() {
        val bankingData = view.receiveData()
        if (EmptyUtil.isNotNull(bankingData)) {
            val mobile = bankingData!!.config.mobile
            if (EmptyUtil.isNotNull(mobile) && EmptyUtil.isNotEmpty(mobile) && ValidationUtil.isMobileNumberValid(mobile)) {
                view.setMobile(bankingData.config.mobile!!)
            }
            view.setBankData(bankingData.bankLogo, bankingData.bankName, bankingData.bankIcon)
            view.setButtonText("Pay Rs " + StringUtil.formatNumber(NumberUtil.convertToRupees(bankingData.config.amount)))

            val clickMap = view.setOnClickListener()
            if (EmptyUtil.isNotNull(clickMap["pay"])) {
                compositeSignal.add(clickMap.getValue("pay")
                        .connect {
                            onFormSubmitted(view.contactNumber, bankingData.bankIdx,
                                    bankingData.bankName, bankingData.config)
                        })
            }
            compositeSignal.add(view.setEditTextListener()
                    .connect {
                        view.setEditTextError(null)
                    })
        }
    }

    override fun onDestroy() {
        compositeSignal.clear()
    }

    override fun onFormSubmitted(mobile: String, bankId: String, bankName: String, config: Config) {
        if (EmptyUtil.isNotEmpty(mobile) && ValidationUtil.isMobileNumberValid(mobile)) {
            if (view.isNetworkAvailable) {
                val map = HashMap<String, String>()
                map["mobile"] = mobile
                map["bankId"] = bankId
                map["bankName"] = bankName
                map["checkout_version"] = BuildConfig.VERSION_NAME
                map["checkout_android_version"] = AppUtil.getOsVersion()
                map["checkout_android_api_level"] = AppUtil.getApiLevel().toString()
                map["public_key"] = config.publicKey
                map["product_identity"] = config.productId
                map["product_name"] = config.productName
                map["amount"] = config.amount.toString()
                map["bank"] = map["bankId"].orEmpty()
                map["source"] = "android"
                map["return_url"] = view.packageName

                if (EmptyUtil.isNotNull(config.productUrl) && EmptyUtil.isNotEmpty(config.productUrl)) {
                    map["product_url"] = config.productUrl!!
                }

                val data = EncodeUtil.urlEncode(map)

//                view.saveConfigInFile(config)
                view.openConfigService()
                view.openEBanking(Constant.url + "ebanking/initiate/?" + data)
                LogUtil.log("data", data)

            } else {
                view.showNetworkError()
            }
        } else {
            if (EmptyUtil.isEmpty(mobile)) {
                view.setEditTextError(EMPTY_ERROR)
            } else {
                view.setEditTextError(MOBILE_ERROR)
            }
        }
    }
}